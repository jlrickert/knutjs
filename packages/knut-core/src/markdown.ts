import { fromMarkdown } from 'mdast-util-from-markdown';
import { toMarkdown } from 'mdast-util-to-markdown';
import { gfmFromMarkdown, gfmToMarkdown } from 'mdast-util-gfm';
import { gfm } from 'micromark-extension-gfm';
import * as Mdast from 'mdast';
import { DexEntry } from './dex.js';

export class MarkdownAST {
	static isTitleToken(token: Mdast.RootContent): token is Mdast.Heading {
		return token.type === 'heading' && token.depth === 1;
	}

	static from(markdown: string): Mdast.Root {
		return fromMarkdown(markdown, 'utf-8', {
			extensions: [gfm()],
			mdastExtensions: [gfmFromMarkdown()],
		});
	}

	static to(root: Mdast.Nodes): string {
		return toMarkdown(root, { extensions: [gfmToMarkdown()] });
	}

	static text(value: string): Mdast.Text {
		return { type: 'text', value };
	}

	static list(
		children: Mdast.ListItem[],
		options?: { ordered?: boolean; spread?: boolean },
	): Mdast.List {
		return {
			type: 'list',
			children,
			ordered: options?.ordered,
			spread: options?.spread,
		};
	}

	static listItem(children: Mdast.BlockContent[]): Mdast.ListItem {
		return { type: 'listItem', children };
	}

	static paragraph(children: Mdast.PhrasingContent[]): Mdast.Paragraph {
		return { type: 'paragraph', children };
	}

	static nodeLink(node: DexEntry): Mdast.Link {
		const id = node.nodeId;
		return {
			type: 'link',
			url: `../${id}`,
			children: [{ type: 'text', value: node.title }],
		};
	}
}

export class MarkdownDocument {
	md = MarkdownAST.from('');

	addList(items: Mdast.ListItem[]): void {
		if (this.md.children)
			this.md.children.push({
				type: 'list',
				children: items,
			});
	}

	stringify(): string {
		return MarkdownAST.to(this.md);
	}
}
