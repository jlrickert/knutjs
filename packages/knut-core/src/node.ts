import { EventEmitter } from 'events';
import { MetaFile as MetaFile, Tag } from './metaFile.js';
import { stringify } from './utils.js';
import { NodeContent } from './nodeContent.js';
import { KegStorage } from './kegStorage.js';
import { TimeLike } from 'fs';

export class NodeId {
	static parsePath(path: string): NodeId | null {
		const parts = path.split('/');
		for (const part of parts) {
			const id = Number(part);
			if (!Number.isNaN(id)) {
				return new NodeId(id);
			}
		}
		return null;
	}

	constructor(public readonly id: number) {}

	next(): NodeId {
		return new NodeId(Number(this.id) + 1);
	}

	lt(other: NodeId): boolean {
		return Number(this.id) < Number(other.id);
	}

	lte(other: NodeId): boolean {
		return Number(this.id) <= Number(other.id);
	}

	gt(other: NodeId): boolean {
		return Number(this.id) > Number(other.id);
	}

	gte(other: NodeId): boolean {
		return Number(this.id) >= Number(other.id);
	}

	eq(other: NodeId): boolean {
		return this.id === other.id;
	}

	getMetaPath(): string {
		return `${this.id}/meta.yaml`;
	}

	getReadmePath(): string {
		return `${this.id}/README.md`;
	}

	stringify(): string {
		return String(this.id);
	}
}

export type NodeData = {
	content: NodeContent;
	updated: string;
	created: string;
	meta: MetaFile;
};

export type NodeOptions = {
	content: string;
	updated: string;
	created: string;
	meta?: MetaFile;
};

/**
 * Node represents an in memory object containing all data
 * to be consisdered a keg node
 **/
export class KegNode extends EventEmitter {
	static isNode(obj: any): obj is KegNode {
		return obj instanceof KegNode;
	}

	static async fromStorage(
		nodeId: NodeId,
		storage: KegStorage,
	): Promise<KegNode | null> {
		const nodePath = NodeContent.filePath(nodeId);
		const metaPath = MetaFile.filePath(nodeId);
		const rawContent = await storage.read(nodePath);
		if (rawContent === null) {
			return null;
		}
		const yamlData = await storage.read(metaPath);
		const metaData = yamlData
			? MetaFile.fromYAML(yamlData)
			: new MetaFile();
		const stats = await storage.stats(nodePath);
		if (stats === null) {
			return null;
		}
		const content = await NodeContent.fromMarkdown(rawContent);
		const now = new Date();
		const node = new KegNode({
			content,
			updated: stringify(stats?.mtime ?? now),
			created: stringify(stats?.btime ?? now),
			meta: metaData,
		});
		return node;
	}

	static async create(): Promise<KegNode> {
		const now = new Date();
		return new KegNode({
			content: await NodeContent.fromMarkdown(''),
			created: stringify(now),
			updated: stringify(now),
			meta: new MetaFile(),
		});
	}

	static parseNodeId(filepath: string): string | null {
		const parts = filepath.split('/');
		parts.pop();
		const nodeId = parts.pop();
		return nodeId ?? null;
	}

	static metaPath(nodeId: NodeId): string {
		return `${nodeId}/meta.yaml`;
	}

	static async fromContent(options: NodeOptions): Promise<KegNode> {
		const content = await NodeContent.fromMarkdown(options.content);
		return new KegNode({
			content,
			meta: options.meta ?? new MetaFile(),
			updated: options.updated,
			created: options.created,
		});
	}

	private constructor(private data: NodeData) {
		super();
	}

	get title(): string {
		return this.data.content.title ?? '';
	}

	set title(value: string) {
		this.data.content.title = value;
	}

	get content(): string {
		return stringify(this.data.content);
	}

	get meta(): MetaFile {
		return this.data.meta;
	}

	get updated(): string {
		return this.data.updated;
	}

	set updated(value: TimeLike) {
		this.data.updated = stringify(value);
	}

	get created(): string {
		return this.data.created;
	}

	set created(value: TimeLike) {
		this.data.created = stringify(value);
	}

	async updateContent(content: string): Promise<void> {
		this.data.content = await NodeContent.fromMarkdown(content);
		this.data.updated = stringify(new Date());

		this.emit('update', {
			node: this,
		});
	}

	updateMeta(f: MetaFile | ((meta: MetaFile) => void)): void {
		if (f instanceof MetaFile) {
			this.data.meta = f;
		} else {
			f(this.data.meta);
		}
	}

	addTag(tag: Tag): void {
		this.data.meta.addTag(tag);
	}

	getTags(): readonly string[] {
		return this.data.meta.getTags();
	}

	addDate(datetime: string): void {
		this.data.meta.add('date', datetime);
	}
}
