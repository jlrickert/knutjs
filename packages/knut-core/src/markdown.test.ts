import { describe, expect, test } from 'vitest';
import { KegNode } from './node';
import { MarkdownAST as AST } from './markdown';

describe('common markdown operations', () => {
	test('should be able to generate a node link', async () => {
		const node = await KegNode.fromContent({
			updated: '2023-23-03',
			content: '# Example title',
		});
		const ast = AST.nodeLink('43', node);
		const content = AST.to(ast);
		expect(content).toEqual('[Example title](../43)\n');
	});

	test('should be able to generate a list', async () => {
		const nodeList = await Promise.all([
			KegNode.fromContent({
				updated: '2023-23-03',
				content: '# Example title 1',
			}),
			KegNode.fromContent({
				updated: '2023-20-03',
				content: '# Example title 2',
			}),
			KegNode.fromContent({
				updated: '2023-21-03',
				content: '# Example title 3',
			}),
		]);
		const ast = AST.list(
			nodeList.map((node, i) => {
				const date = AST.text(node.updated);
				const space = AST.text(' ');
				const link = AST.nodeLink(String(i), node);
				return AST.listItem([AST.paragraph([date, space, link])]);
			}),
			{ spread: false },
		);
		const content = AST.to(ast);
		expect(content).matchSnapshot('list of items');
	});
});
