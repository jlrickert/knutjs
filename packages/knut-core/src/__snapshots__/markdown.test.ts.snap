// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`common markdown operations > should be able to generate a list > list of items 1`] = `
"* 2023-23-03 [Example title 1](../0)
* 2023-20-03 [Example title 2](../1)
* 2023-21-03 [Example title 3](../2)
"
`;
